### ISSUE TITLE
Fine-tuning: How to fix the weights of pre-trained model in the first couple of layers and tune the last layer only?

### ISSUE THREAD

**Issue by luzhi** [vlfeat/matconvnet#643]
I would like to fine tune the parameters of the last layer FC of VGG-16 model. At the same time, I want to use the weights in the rest layers that were trained in the pre-trained model. Could you please advise me how can I do that in MatConvNet? Thanks!

**Comment by yw155** [comment]
You can set the learning rate and decay parameters of the layers which you do not want to tune to zeros.

**Comment by albanie** [comment]
As @yw155 pointed out, this can be achieved by setting the learning rate of the layers you wish to "freeze" to zero.  When the learning rate is zero, the weight decay is no longer a factor in the gradient updates if you are using standard SGD (which you can see [here](https://github.com/vlfeat/matconvnet/blob/master/examples/cnn_train.m#L417-L418)), but it may be useful to set them to zero if you are using a different optimization technique.  

Here's an example.  The parameter indexing is done one step at a time which hopefully makes things a little clearer at the expense of slightly more code 😉 

    % load simpleNN vgg-16 model (change the file path to wherever your 
    % model is stored)
    net = load('~/data/models/matconvnet/imagenet-vgg-verydeep-16.mat');

    % convert to DagNN
    dagnet = dagnn.DagNN.fromSimpleNN(net);

    % set the name of the target layer we want to fine tune
    layerName = 'fc8';

    % find its layer index in the DagNN
    layerIndex = dagnet.getLayerIndex(layerName);

    % find the parameter names for the fc8 layer
    paramNames = [dagnet.layers(layerIndex).params];

    % find the indices of these paramters 
    paramIdx = dagnet.getParamIndex(paramNames);

    % now we have their indices, we can set their learning rate to the 
    % desired value
    learningRate = 1;
    [dagnet.params(paramIdx).learningRate] = deal(learningRate);

    % similarly, we want to find all the parameters that belong to the layers
    % before the last fully connected layer - we can freeze them by 
    % setting their learning rate to zero
    frozenParamIdx = dagnet.getParamIndex([dagnet.layers(1:end-2).params]);
    frozenRate = 0;
    [dagnet.params(frozenParamIdx).learningRate] = deal(frozenRate);

    % ready to fine tune!

**Comment by luzhi** [comment]
@yw155 @albanie Thank you guys!

**Comment by FrancisYizhang** [comment]
@albanie 
Should the weight decay is set to 0?
